### ISSUE TITLE
Pathfinding bug when using large tokens

### ISSUE THREAD

**Issue by Piglit** [RPTools/maptool#2409]
**Describe the bug**
When using large tokens, there is a bug with the automated Pathfinding when moving left to right touching a obstacle just north of the path. In some cases not the shortes path is found, sometimes preventing walking thorugh a narrow corridor.

**To Reproduce**
* Create a token, set the size to "large"
* create a VBL/MBL (grid aligned)
* Make sure the "AI"-Pathfinding is active.
* Drag the token from left to right just south of the VBL.
* The bug does not always occure, and it occures only moving from left to right, never when moving in other directions.
* I only tested with "NO_DIAGONALS" metric on a sqare grid.
* When the bug occures, it does always occure on the same position. However I can see no predictable way on which position it will occure.

**Expected behavior**
I would expect the shortest path is found (like when moving right to left).

**Screenshots**

Not the shorest path is found:
![simple](https://user-images.githubusercontent.com/34441046/108621761-ebe97880-7434-11eb-89ef-68d4f38a9a90.png)

No way though the passage is found at all:
![narrow](https://user-images.githubusercontent.com/34441046/108621766-f0159600-7434-11eb-953a-f49aaead5d55.png)

But it works when walking in the opposite direction:
![right_to_left](https://user-images.githubusercontent.com/34441046/108621769-f1df5980-7434-11eb-8dfe-302e4df132e4.png)

**MapTool Info**
- Version: 1.7.0 and 1.8.3
- Install: Upgraded since 1.6.x

**Desktop (please complete the following information):**
 - OS: Windows
 - Version 10

**Comment by kwvanderlinde** [comment]
I happened to address this in #2721 as the same underlying issue would have caused trouble for all token sizes under the new 1-2-1 implementation.

**Comment by Phergus** [comment]
There are still some hiccups with tokens larger than a single cell that are caused by using the upper-left cell for determining movement.
![image](https://user-images.githubusercontent.com/34379254/121435033-7a79fb00-c93b-11eb-962d-d8a61c58c55c.png)
The large troll token can partially overlap the wall by one cell.  The huge dragon can do so by two cells.
Or if dragged against the bottom of the open area.
![image](https://user-images.githubusercontent.com/34379254/121435342-f116f880-c93b-11eb-9561-b9e670cf3ee4.png)
Both can overlap by a full row.

On a map with FoW, if a player were to end the move with the large troll token's partially in the wall, this results in the token being lost from Player's view.
![image](https://user-images.githubusercontent.com/34379254/121436121-37208c00-c93d-11eb-94e3-0afa8ba4df84.png)
The huge dragon token can only move into the wall by one and so is still visible.  (For those games where players tokens can be large or greater in size.)

**Comment by kwvanderlinde** [comment]
I'll revist this later today hopefully. I'm not 100% sure right now, but I think this is a simple matter of checking whether a given step is blocked for each occupied cell. I.e., checking whether the bottom-right corner would be blocked from moving to where the new position of bottom-right corner would be.

The same or similar behaviour can probably be seen when using tokens to block movement.

**Comment by Phergus** [comment]
> The same or similar behaviour can probably be seen when using tokens to block movement.
![image](https://user-images.githubusercontent.com/34379254/121439793-dcd6f980-c943-11eb-8776-29ebf7ec2142.png)
Yup.  The T's are terrain mod tokens.

**Comment by kwvanderlinde** [comment]
Now that I've had a chance to look at the issue closer, I see there are two different aspects that need to be handled if we want this fixed robustly:
1. For any given move from one cell to a neighboring cell, check whether the move is blocked by MBL. The code currently mishandles this, leading to the behaviour shown. My idea above will fix it.
2. Check whether a position is allowed by MBL. My idea has no impact on this, which causes other odd situations. E.g., a large token is allowed to move through a line of MBL, provided the line can "slip between" the cells.

To handle (2), what we can do is build geometry for the entire token footprint (minus half a cell around the boundary), and then check whether MBL intersects it. If so, we know the position is invalid and can be closed. Otherwise, we can go forward with checking (1) and the other details around neighbor selection.

Fixing blocking terrain isn't too difficult either, but I'm going to suggest it gets handled in a separate issue, as it relates to a more general problem of how to handle multiple terrain modifiers with large tokens, and how that compares to current behaviour.

**Comment by Phergus** [comment]
Yeah a separate issue for dealing with terrain blocking is a good idea.

Do/should terrain modifiers apply to larger tokens?  Better pursued in the separate issue as well I reckon.

**Comment by Phergus** [comment]
Tested latest code.  Large token movement through constrained spaces generally improved.

This case had worse results.  Starting with this move:
![image](https://user-images.githubusercontent.com/34379254/121707590-f0917580-ca93-11eb-8c8e-6ed01c29591a.png)
And then shifting up one:
![image](https://user-images.githubusercontent.com/34379254/121707566-ea02fe00-ca93-11eb-85b7-d00617d6fdba.png)
Takes about 8 secs before the pathfinder gives up and gives a 0 movement cost.  Was about 3 secs under 1.9.0-b2.

**Comment by kwvanderlinde** [comment]
Hmm, that's an odd discrepancy, and I can't reproduce it. "Giving up" means timing out, which should always be after 5 seconds for a GM, or 10 seconds for a player. The only other way for the time to be significantly different is for one step of A* to take a really long time (~5 second in this case), and I don't see how that could be happening.

Could you share your map with me? Just in case there's some subtlety causing this behaviour.

**Comment by Phergus** [comment]
Will do.  This was on a player client.  Doing most testing there as GMs can move tokens anywhere.

Remove `.zip`.
[PathfindingTesting.rpmap.zip](https://github.com/RPTools/maptool/files/6639932/PathfindingTesting.rpmap.zip)

**Comment by kwvanderlinde** [comment]
I've tried it out and can't reproduce it. On 1.9.0-b2 I get the same results as on develop: timeout after 5s on GM client, and timeout after 10s on player client.

Are you sure the 3s case was done on a player client? This example should be running into #2355, where A* terminates only on timeout, and that should be consistent.

**Comment by Phergus** [comment]
Quite possible I was using the GM client when the short timeout occurred.  Will retest.

**Comment by Phergus** [comment]
Retested and the timeout was 5s for GM and 10s for player under beta 2.  All good.
